client/src/components/Workflow/Editor/Forms/FormDefault.vue

```diff
@@ -0,0 +1,139 @@
+                @click="onUpgradeSubworkflow"
+            >
+                <span class="fa fa-sync" />
+            </b-button>
+        </template>
+        <template v-slot:body>
+            <FormElement
+                id="__label"
+                :value="node.label"
+                title="Label"
+                help="Add a step label."
+                @onChange="onLabel"
+                :error="errorLabel"
+            />
+            <FormElement
+                id="__annotation"
+                :value="node.annotation"
+                title="Step Annotation"
+                :area="true"
+                help="Add an annotation or notes to this step. Annotations are available when a workflow is viewed."
+                @onChange="onAnnotation"
+            />
+            <Form :id="id" :inputs="inputs" @onChange="onChange" />
+        </template>
+    </FormCard>
+</template>
+
+<script>
+import axios from "axios";
+import { getAppRoot } from "onload/loadConfig";
+import Form from "components/Form/Form";
+import FormCard from "components/Form/FormCard";
+import FormElement from "components/Form/FormElement";
+import { checkLabels } from "components/Workflow/Editor/modules/utilities";
+import WorkflowIcons from "components/Workflow/icons";
+
+export default {
+    components: {
+        Form,
+        FormCard,
+        FormElement,
+    },
+    props: {
+        datatypes: {
+            type: Array,
+            required: true,
+        },
+        getManager: {
+            type: Function,
+            required: true,
+        },
+        getNode: {
+            type: Function,
+            required: true,
+        },
+    },
+    data() {
+        return {
+            errorLabel: null,
+        };
+    },
+    computed: {
+        node() {
+            return this.getNode();
+        },
+        nodeIcon() {
+            return WorkflowIcons[this.node.type];
+        },
+        workflow() {
+            return this.getManager();
+        },
+        id() {
+            return this.node.id;
+        },
+        inputs() {
+            return this.node.config_form.inputs;
+        },
+        isSubworkflow() {
+            return this.node.type == "subworkflow";
+        },
+    },
+    methods: {
+        onAnnotation(newAnnotation) {
+            this.node.setAnnotation(newAnnotation);
+        },
+        onLabel(newLabel) {
+            this.node.setLabel(newLabel);
+            this.errorLabel = checkLabels(this.node.id, newLabel, this.workflow.nodes);
+        },
+        onEditSubworkflow() {
+            this.workflow.onEditSubworkflow(this.node.content_id);
+        },
+        onUpgradeSubworkflow() {
+            this.workflow.onAttemptRefactor([
+                { action_type: "upgrade_subworkflow", step: { order_index: parseInt(this.node.id) } },
+            ]);
+        },
+        onChange(values) {
+            axios
+                .post(`${getAppRoot()}api/workflows/build_module`, {
```

Comment:
Shouldn't the onChange event be passed up to the Node ?
Manipulating another Vue app like this seems like a pattern we want to avoid ?

Reply:
Yes this one should be changed too in the near future but currently it should probably stay here since it is inter winded with calling form update handlers. Once the underlying regular form wrapper has progressed we should be able to emit an event here too.

Reply:
Thank you for looking into this!